pkg/devfile/generator/policy.go

```diff
@@ -0,0 +1,131 @@
+	}
+}
+
+func patchForPolicy(podTemplateSpec *corev1.PodTemplateSpec, policy psaapi.Policy) (*corev1.PodTemplateSpec, error) {
+	evaluator, err := psapolicy.NewEvaluator(psapolicy.DefaultChecks())
+	if err != nil {
+		return nil, err
+	}
+	results := evaluator.EvaluatePod(policy.Enforce, &podTemplateSpec.ObjectMeta, &podTemplateSpec.Spec)
+	for _, result := range results {
+		if result.Allowed {
+			continue
+		}
+		switch result.ForbiddenReason {
+		case "allowPrivilegeEscalation != false":
+			podTemplateSpec = patchAllowPrivilegeEscalation(podTemplateSpec)
+		case "unrestricted capabilities":
+			podTemplateSpec = patchUnrestrictedCapabilities(podTemplateSpec)
+		case "runAsNonRoot != true":
+			podTemplateSpec = patchRunAsNonRoot(podTemplateSpec)
+		case "seccompProfile":
+			podTemplateSpec = patchSeccompProfile(podTemplateSpec, policy.Enforce.Level)
+			// Other policies are not implemented as they cannot be encountered with pods created by the library
+		}
+	}
+
+	newResults := evaluator.EvaluatePod(policy.Enforce, &podTemplateSpec.ObjectMeta, &podTemplateSpec.Spec)
+	for _, result := range newResults {
+		if !result.Allowed {
+			// This is an assertion for future developers, and should never happen in production
+			// This could happen during development if some unsecure fields are added from `getPodTemplateSpec` or `GetContainers`/`GetInitContainers`
+			return nil, fmt.Errorf("error patching pod for Pod Security Admission. The folowing policy is still not respected: %s (%s)", result.ForbiddenReason, result.ForbiddenDetail)
+		}
+	}
+
+	return podTemplateSpec, nil
+}
+
+func patchAllowPrivilegeEscalation(podTemplateSpec *corev1.PodTemplateSpec) *corev1.PodTemplateSpec {
+	visitContainers(podTemplateSpec, func(container *corev1.Container) {
+		if container.SecurityContext == nil {
+			container.SecurityContext = &corev1.SecurityContext{}
+		}
+		container.SecurityContext.AllowPrivilegeEscalation = pointer.Bool(false)
+	})
+	return podTemplateSpec
+}
+
+func patchUnrestrictedCapabilities(podTemplateSpec *corev1.PodTemplateSpec) *corev1.PodTemplateSpec {
+	visitContainers(podTemplateSpec, func(container *corev1.Container) {
+		if container.SecurityContext == nil {
+			container.SecurityContext = &corev1.SecurityContext{}
+		}
+		if container.SecurityContext.Capabilities == nil {
+			container.SecurityContext.Capabilities = &corev1.Capabilities{}
+		}
+		container.SecurityContext.Capabilities.Drop = append(container.SecurityContext.Capabilities.Drop, "ALL")
+	})
+	return podTemplateSpec
+}
+
+func patchRunAsNonRoot(podTemplateSpec *corev1.PodTemplateSpec) *corev1.PodTemplateSpec {
+	if podTemplateSpec.Spec.SecurityContext == nil {
+		podTemplateSpec.Spec.SecurityContext = &corev1.PodSecurityContext{}
+	}
+	podTemplateSpec.Spec.SecurityContext.RunAsNonRoot = pointer.Bool(true)
+	// No need to set the value as true for containers, as setting at the Pod level is sufficient
```

Comment:
To me (and per https://kubernetes.io/docs/tasks/configure-pod-container/security-context/#set-the-security-context-for-a-container), security settings at the Container level override settings made at the Pod level when there is an overlap. So I think, we don't need to set the value as `true` to containers only if they are not specified at the container level.
So we should still visit all containers and set the field to `true`, no?

Reply:
> To me (and per https://kubernetes.io/docs/tasks/configure-pod-container/security-context/#set-the-security-context-for-a-container), security settings at the Container level override settings made at the Pod level when there is an overlap. So I think, we don't need to set the value as `true` to containers only if they are not specified at the container level. 

I agree up to here.

> So we should still visit all containers and set the field to `true`, no?

By design, we know the field at container level will never be set (because they could be set only from the library, and they are not). For this reason, we only need to set it at pod level.